lms/djangoapps/grades/api/v1/views.py

```diff
@@ -0,0 +1,233 @@
+from student.roles import CourseStaffRole
+from util.string_utils import str_to_bool
+
+log = logging.getLogger(__name__)
+USER_MODEL = get_user_model()
+
+
+@view_auth_classes()
+class GradeViewMixin(DeveloperErrorViewMixin):
+    """
+    Mixin class for Grades related views.
+    """
+    def _get_course(self, course_key_string, user, access_action):
+        """
+        Returns the course for the given course_key_string after
+        verifying the requested access to the course by the given user.
+        """
+        try:
+            course_key = CourseKey.from_string(course_key_string)
+        except InvalidKeyError:
+            return self.make_error_response(
+                status_code=status.HTTP_404_NOT_FOUND,
+                developer_message='The provided course key cannot be parsed.',
+                error_code='invalid_course_key'
+            )
+
+        try:
+            return courses.get_course_with_access(
+                user,
+                access_action,
+                course_key,
+                check_if_enrolled=True,
+            )
+        except Http404:
+            log.info('Course with ID "%s" not found', course_key_string)
+        except CourseAccessRedirect:
+            log.info('User %s does not have access to course with ID "%s"', user.username, course_key_string)
+        return self.make_error_response(
+            status_code=status.HTTP_404_NOT_FOUND,
+            developer_message='The user, the course or both do not exist.',
+            error_code='user_or_course_does_not_exist',
+        )
+
+    def _get_effective_user(self, request, course):
+        """
+        Returns the user object corresponding to the request's 'username' parameter,
+        or the current request.user if no 'username' was provided.
+
+        Verifies that the request.user has access to the requested users's grades.
+        Returns a 403 error response if access is denied, or a 404 error response if the user does not exist.
+        """
+
+        # Use the request user's if none provided.
+        if 'username' in request.GET:
+            username = request.GET.get('username')
+        else:
+            username = request.user.username
+
+        if request.user.username == username:
+            # Any user may request her own grades
+            return request.user
+
+        # Only a user with staff access may request grades for a user other than herself.
+        if not has_access(request.user, CourseStaffRole.ROLE, course):
+            log.info(
+                'User %s tried to access the grade for user %s.',
+                request.user.username,
+                username
+            )
+            return self.make_error_response(
+                status_code=status.HTTP_403_FORBIDDEN,
+                developer_message='The user requested does not match the logged in user.',
+                error_code='user_mismatch'
+            )
```

Comment:
Instead of the above code from L72-92, use the [IsUserInUrlOrStaff Mixin](https://github.com/edx/edx-platform/blob/e8469a45aaddfa83fe671978cb6230744b2524c3/openedx/core/lib/api/permissions.py#L128).

Reply:
This mixin doesn't seem to apply to course staff roles. The current implementation allows course or global staff to request the course. Is this not intended? Should it just be global staff?

Reply:
I think it's better to be consistent with the behavior of the other endpoints right now, which, yes, do require global staff access.
In the subsequent PR this will be improved once we use scopes.